llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp

```diff
@@ -4540,14 +4540,45 @@ static Instruction *foldICmpXorXX(ICmpInst &I, const SimplifyQuery &Q,
   if (!match(Op0, m_c_Xor(m_Specific(Op1), m_Value(A))))
     return nullptr;
 
+  CmpInst::Predicate StrictPred = ICmpInst::getStrictPredicate(Pred);
+  // Transform
+  // X s< ~X, X s<= ~X, X u> ~X, X u>= ~X
+  // --> X s< 0
+  // X s> ~X, X s>= ~X, X u< ~X, X u<= ~X
+  // --> X s> -1
+  // X ==  ~X  -->   false
+  // X !=  ~X  -->   true
+  if (match(Op0, m_Not(m_Specific(Op1)))) {
```

Comment:
These all work for any `xor x, neg_C`:
https://alive2.llvm.org/ce/z/6LR8Nn

Reply:
(or any negative value in general).

Reply:
This is still outstanding (you don't need to use `isKnownNegative`, but we should any negative constant at least).

Reply:
at first when I implement this pattern, I wrote code as comparing sign-bit between them. so, it seems can be comparing between x and neg_x also can be use this pattern.

Reply:
not x and neg_x, but `x` and `x ^ neg_c` (although the eq/ne ones work for any `x` and `x ^ NonZero`.

Reply:
I understand now. there is need to improve our code to use `isKnownNegative` in the future?

Reply:
I don't think so, when I initially submitted this function I think i had that logic in here, but @nikic was opposed. If it comes up in real code we can re-evaluate.